hypothesis-python/src/hypothesis/core.py

```diff
@@ -398,72 +399,38 @@ def process_arguments_to_given(
     return arguments, kwargs, test_runner, search_strategy
 
 
-def run_once(fn):
-    """Wraps a no-args function so that its outcome is cached.
-    We use this for calculating various lists of exceptions
-    the first time we use them."""
-    result = [None]
-
-    def run():
-        if result[0] is None:
-            result[0] = fn()
-            assert result[0] is not None
-        return result[0]
-
-    run.__name__ = fn.__name__
-    return run
-
-
-@run_once
 def skip_exceptions_to_reraise():
     """Return a tuple of exceptions meaning 'skip this test', to re-raise.
 
     This is intended to cover most common test runners; if you would
     like another to be added please open an issue or pull request.
     """
-    import unittest
-
     # This is a set because nose may simply re-export unittest.SkipTest
-    exceptions = {unittest.SkipTest}
-
-    try:  # pragma: no cover
-        from unittest2 import SkipTest
-
-        exceptions.add(SkipTest)
-    except ImportError:
-        pass
-
-    try:  # pragma: no cover
-        from pytest.runner import Skipped
-
-        exceptions.add(Skipped)
-    except ImportError:
-        pass
-
-    try:  # pragma: no cover
-        from nose import SkipTest as NoseSkipTest
-
-        exceptions.add(NoseSkipTest)
-    except ImportError:
-        pass
-
+    exceptions = set()
+    # We use this sys.modules trick to avoid importing libraries -
+    # you can't be an instance of a type from an unimported module!
+    # This is fast enough that we don't need to cache the result,
+    # and more importantly it avoids possible side-effects :-)
+    if "unittest" in sys.modules:
+        exceptions.add(sys.modules["unittest"].SkipTest)
+    if "unittest2" in sys.modules:  # pragma: no cover
+        exceptions.add(sys.modules["unittest2"].SkipTest)
+    if "nose" in sys.modules:  # pragma: no cover
+        exceptions.add(sys.modules["nose"].SkipTest)
+    if "_pytest" in sys.modules:  # pragma: no branch
+        exceptions.add(sys.modules["_pytest"].outcomes.Skipped)
     return tuple(sorted(exceptions, key=str))
 
 
-@run_once
 def failure_exceptions_to_catch():
```

Comment:
I am not wild about running these methods on every exception raised - given that that raising exceptions is the main thing Hypothesis tests do it's pretty hot path, and doing all this conditional querying as to whether stuff is already in `sys.modules` is technically fine but looks pretty weird and results in all of these coverage pragmas.

How about instead catching all exceptions and checking whether the current exception is one to reraise based on a string version of its name? e.g. something like:

```python
from hypothesis.internal.compat import qualname

SKIP_EXCEPTIONS = frozenset({
    'unittest.SkipTest', 
    'unittest2.SkipTest',
    'nose.SkipTest',
    '_pytest.outcomes.Skipped', 
})

...

except BaseException as e:
    if qualname(e) in SKIP_EXCEPTIONS or (
        not isinstance(e, Exception) and qualname(e) != '_pytest.outcomes.Failed'
    ):
        raise
    # normal handling logic continues here
    
```

Reply:
That really doesn't look better to me! The sys.modules checks are unusual, but the whole function is fast - we have four dict lookups, and up to five attribute accesses. The slowest part is sorting the types by str!

And we already had those coverage pragmas, spread out across more lines of code.